### PR TITLE
YouTube Tracking

### PR DESCRIPTION
**Pull request by markjamesbutler**
## What does this change?

Adds GA and Ophan tracking to You Tube atoms.

## What is the value of this and can you measure success?

None really except parity with existing video.js tracking (except pre-roll events which are not supported.)

## Does this affect other platforms - Amp, Apps, etc?

No

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/video/events.js [hunk 1]
```diff
@@ -119,25 +121,6 @@ define([
         return action;
     }
 
-    function buildGoogleAnalyticsEvent(mediaEvent, metrics, canonicalUrl) {
```
Comment (akash1810) [inline static/src/javascripts/projects/common/modules/video/events.js]: should `getGoogleAnalyticsEventAction` go into the `ga-helper` module too?

File: static/src/javascripts/projects/common/modules/video/events.js [hunk 2]
```diff
@@ -152,7 +135,8 @@ define([
             return 'media:' + eventName;
         }).forEach(function(playerEvent) {
             player.on(playerEvent, function(_, mediaEvent) {
-                ga(gaTracker + '.send', 'event', buildGoogleAnalyticsEvent(mediaEvent, events, canonicalUrl));
+                ga(gaTracker + '.send', 'event', gaHelper.buildGoogleAnalyticsEvent(mediaEvent, events, canonicalUrl,
+                    'guardian-videojs', getGoogleAnalyticsEventAction, mediaEvent.mediaId));
```
Comment (akash1810) [inline static/src/javascripts/projects/common/modules/video/events.js]: is `mediaEvent.mediaId` the atom id or yt video id?

File: static/src/javascripts/projects/common/modules/video/youtube-player.js [hunk 3]
```diff
@@ -2,19 +2,42 @@ define([
...
     });
 
+    var players = {},
+        progressTracker = {};
+
+    var STATES = {
+        'ENDED': _onPlayerEnded,
+        'PLAYING': _onPlayerPlaying,
+        'PAUSED': _onPlayerPaused,
+        'BUFFERING': null,
```
Comment (akash1810) [inline static/src/javascripts/projects/common/modules/video/youtube-player.js]: why are we mapping `BUFFERING` and `CUED`?

File: static/src/javascripts/projects/common/modules/video/youtube-player.js [hunk 4]
```diff
@@ -56,31 +92,81 @@ define([
...
         });
     }
 
+    function recordPlayerProgress(id, atomId) {
+
+        var currentTime = players[id].player.getCurrentTime(),
+            percentPlayed = Math.round(((currentTime / players[id].duration) * 100));
+
+        if (percentPlayed > 0 &&
+            percentPlayed % 25 === 0) {
```
Comment (akash1810) [inline static/src/javascripts/projects/common/modules/video/youtube-player.js]: are we firing a `100%` and an `end` event then?

File: static/src/javascripts/projects/common/modules/video/youtube-tracking.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                'play': 'metric1',
+                'skip': 'metric2',
+                '25': 'metric3',
+                '50': 'metric4',
+                '75': 'metric5',
+                'end': 'metric6'
+            },
+            baseEventObject: {
+                eventCategory: 'Media',
+                eventAction: 'video content',
```
Comment (akash1810) [inline static/src/javascripts/projects/common/modules/video/youtube-tracking.js]: can we reuse `eventAction` from above?

File: static/src/javascripts/projects/common/modules/video/youtube-tracking.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                eventLabel: window.location.pathname,
```
Comment (akash1810) [inline static/src/javascripts/projects/common/modules/video/youtube-tracking.js]: we'd want this to be the atom id, to identify each individual atom rather than just the page I think

File: static/src/javascripts/projects/common/modules/video/youtube-tracking.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            }
+        };
+
+        var eventsList = ['play', '25', '50', '75', 'end'];
+
+        forEach(eventsList, function(event) {
+            mediator.once(buildEventId(event, videoId), function(id) {
+                ophanRecord(event, id);
+                ga(gaTracker + '.send', 'event',
+                    gaHelper.buildGoogleAnalyticsEvent(event, events, window.location.pathname,
```
Comment (akash1810) [inline static/src/javascripts/projects/common/modules/video/youtube-tracking.js]: same as above re `window.location.pathname`

File: static/src/javascripts/projects/common/modules/video/ga-helper.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
+define([
+], function (
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/video/ga-helper.js]: There's no need for the first parameter if the dependency array is empty

File: static/src/javascripts/projects/common/modules/video/events.js (hunk 1 above)
Comment (markjamesbutler) [inline static/src/javascripts/projects/common/modules/video/events.js]: @akash1810 I didn't move this as it isn't shared between the events.js implementation and the new you-tube-tracking.js implemenation.

File: static/src/javascripts/projects/common/modules/video/events.js (hunk 2 above)
Comment (markjamesbutler) [inline static/src/javascripts/projects/common/modules/video/events.js]: It depends which implementation you are using.  The hosted implementation uses youtube_id our implementation users atom id.

From hosted on production: 

https://ophan.theguardian.com/a.gif?viewId=ivdvgr9tla1cr8wvdnbe&video={"id":"gu-video-youtube-aR5_bhSO_d0","eventType":"video:content:play"}

File: static/src/javascripts/projects/common/modules/video/youtube-tracking.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    function buildEventId(event, videoId) {
+        return event + ':' + videoId;
+    }
+
+    function initYoutubeEvents(videoId) {
+
+        var ga = window.ga,
+            gaTracker = config.googleAnalytics.trackers.editorial;
```
Comment (dominickendrick) [inline static/src/javascripts/projects/common/modules/video/youtube-tracking.js]: This is a tiny nit, but can you have a declaration for each `var` rather that comma separate them.

File: static/src/javascripts/projects/common/modules/video/events.js (hunk 2 above)
Comment (akash1810) [inline static/src/javascripts/projects/common/modules/video/events.js]: The hosted pages use atoms too iirc, worth checking with them how they want to track it. My vote would be to use atom id everywhere for simplicity.

cc @guardian/labs-beta @guardian/labs-beta)

File: static/src/javascripts/projects/common/modules/video/events.js (hunk 1 above)
Comment (akash1810) [inline static/src/javascripts/projects/common/modules/video/events.js]: but its a helper module 😉. As far as I can tell its only used there too (albeit passed in as an argument).

File: static/src/javascripts/projects/common/modules/video/events.js (hunk 2 above)
Comment (markjamesbutler) [inline static/src/javascripts/projects/common/modules/video/events.js]: @kelvin-chappell I agree but let's make this change in a separate PR as I don't want to change the PROD behaviour for hosted in this PR.

File: static/src/javascripts/projects/common/modules/video/youtube-tracking.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            }
+        };
+
+        var eventsList = ['play', '25', '50', '75', 'end'];
+
+        forEach(eventsList, function(event) {
+            mediator.once(buildEventId(event, videoId), function(id) {
+                ophanRecord(event, id);
+                ga(gaTracker + '.send', 'event',
+                    gaHelper.buildGoogleAnalyticsEvent(event, events, videoId,
```
Comment (markjamesbutler) [inline static/src/javascripts/projects/common/modules/video/youtube-tracking.js]: Yes, I fixed this yesterday as part of the custom metric change.  It's working as expected now.

**Comment by akash1810** [comment]
lgtm 👍

File: static/src/javascripts/bootstraps/enhanced/main.js [hunk 11]
```diff
@@ -162,6 +162,14 @@ define([
             });
         }
 
+        fastdom.read(function() {
+            if ( $('.atom--media--youtube').length > 0) {
```
Comment (akash1810) [inline static/src/javascripts/bootstraps/enhanced/main.js]: ❤️

File: static/src/javascripts/bootstraps/enhanced/youtube.js [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        'PAUSED': _onPlayerPaused
+    };
+
+    var progressTracker = {};
+    var players = {};
+
+    function init() {
+
+        fastdom.read(function () {
+                        $('.atom--media--youtube').each(function (el) {
```
Comment (akash1810) [inline static/src/javascripts/bootstraps/enhanced/youtube.js]: extreme indenting...

File: static/src/javascripts/bootstraps/enhanced/youtube.js [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+                            var atomId = el.getAttribute('data-media-atom-id');
+                            var videoId = el.firstElementChild.id;
+                            tracking.init(atomId);
```
Comment (akash1810) [inline static/src/javascripts/bootstraps/enhanced/youtube.js]: is this right? `init` seems to want a video id - https://github.com/guardian/frontend/pull/14965/files#diff-76e86efc7805cf2a7c9429d5276bf49cR86

File: static/src/javascripts/projects/common/modules/video/youtube-tracking.js [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+                'skip': 'metric2',
+                '25': 'metric3',
+                '50': 'metric4',
+                '75': 'metric5',
+                'end': 'metric6'
+            },
+            baseEventObject: {
+                eventCategory: 'Media',
+                eventAction: eventAction(),
+                eventLabel: videoId,
```
Comment (akash1810) [inline static/src/javascripts/projects/common/modules/video/youtube-tracking.js]: should this be the atom id?

File: static/src/javascripts/bootstraps/enhanced/youtube.js [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+                                        // Record the duration for percentage calculation.
+                                        player.duration = player.getDuration();
+                                    }
+                    }
+                    , videoId);
+            });
+        });
+    }
+
+    function checkState(id, atomId, state, player, status) {
```
Comment (SiAdcock) [inline static/src/javascripts/bootstraps/enhanced/youtube.js]: The following functions are only used in the context of the `onPlayerStateChange` handler. Perhaps these functions and the `STATES` object should be defined inside the `onPlayerStateChange` handler so we don't have to pass the `id`, `atomId` and `player` around quite so much?

File: static/src/javascripts/bootstraps/enhanced/youtube.js [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+        if (state === window.YT.PlayerState[status] && STATES[status]) {
+            STATES[status](id, atomId, player);
+        }
+    }
+
+    function _onPlayerPlaying(id, atomId, player) {
```
Comment (SiAdcock) [inline static/src/javascripts/bootstraps/enhanced/youtube.js]: Why do some of the function names start with `_` and some do not? All the functions are effectively private within this module.

File: static/src/javascripts/bootstraps/enhanced/youtube.js [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+                            tracking.init(atomId);
+                            players[videoId] = youtubePlayer.init(el,
+                                {
+                                    onPlayerStateChange: function (event) {
+                                        Object.keys(STATES).forEach(checkState.bind(null, videoId, atomId, event.data, event.target));
+                                    },
+                                    onPlayerReady: function (event) {
+                                        var player = event.target;
+                                        // Record the duration for percentage calculation.
+                                        player.duration = player.getDuration();
```
Comment (SiAdcock) [inline static/src/javascripts/bootstraps/enhanced/youtube.js]: Is this useful? We only access this `duration` property once from within the `recordPlayerProgress` function below, but we could just as easily call `getDuration()` there instead

File: static/src/javascripts/bootstraps/enhanced/youtube.js (hunk 13 above)
Comment (markjamesbutler) [inline static/src/javascripts/bootstraps/enhanced/youtube.js]: @akash1810 This is because hosted uses youtube-id's whereas we use atom-id's.  Hence the generic video-id identifier.  I agreed to make the hosted switch to atom-id's in a separate PR.  I've updated the videoID identifier in youtube-player to be named youtube-id to be more explicit.

File: static/src/javascripts/projects/common/modules/video/youtube-tracking.js (hunk 14 above)
Comment (markjamesbutler) [inline static/src/javascripts/projects/common/modules/video/youtube-tracking.js]: @akash1810 See above comment.  VideoID is generic depending on underlying implementation.  i.e. It could support any type of videoId (youtube, atom, future id's?)

File: static/src/javascripts/bootstraps/enhanced/youtube.js (hunk 15 above)
Comment (markjamesbutler) [inline static/src/javascripts/bootstraps/enhanced/youtube.js]: 💯

File: static/src/javascripts/bootstraps/enhanced/youtube.js (hunk 13 above)
Comment (akash1810) [inline static/src/javascripts/bootstraps/enhanced/youtube.js]: shoot, sorry yeah forgot about that

File: static/src/javascripts/projects/common/modules/video/youtube-tracking.js (hunk 14 above)
Comment (akash1810) [inline static/src/javascripts/projects/common/modules/video/youtube-tracking.js]: could we rename the variable to be more generic in that case, and maybe add a comment?

**Comment by SiAdcock** [comment]
👍
